src/datetime.rs

```diff
@@ -1,12 +1,63 @@
 use std::{
-    convert::TryInto,
+    convert::{TryFrom, TryInto},
     error,
     fmt::{self, Display},
     result,
     time::{Duration, SystemTime},
 };
 
-use time::format_description::well_known::Rfc3339;
+use time::{format_description::well_known::Rfc3339, Date};
+use typed_builder::TypedBuilder;
+
+/// Struct providing a builder to construct a `bson::DateTime` from a given year, month, and
+/// date, and optionally the hour, minute, second and millisecond, which default to 0 if not
+/// explicitly set.
+///
+/// ```
+/// use bson::{DateTime, datetime::RawDateTime};
+/// use std::error::Error;
+///
+/// fn main() -> Result<(), Box<dyn Error>> {
+///     let dt = RawDateTime::builder().year(1998).month(2).day(12).minute(1).millisecond(23).build().to_bson()?;
+///     let expected = DateTime::parse_rfc3339_str("1998-02-12T00:01:00.023Z")?;
+///     assert_eq!(dt, expected);
+///     Ok(())
+/// }
+/// ```
+#[derive(TypedBuilder)]
+pub struct RawDateTime {
```

Comment:
Oh hmm, this is a bit of an issue here. I don't think we necessarily want to be introducing a new date time type just for this method, but `TypedBuilder` won't let us customize the `build` method to produce something other than what we're deriving the builder from (see https://github.com/idanarye/rust-typed-builder/issues/52).

Do you still have the manual implementation laying around? We may want to use that actually :grimacing: (sorry!)

What do you guys think @isabelatkinson @abr-egn? Any ideas on how we could leverage `TypedBuilder` here without making a new datetime type?

Reply:
I was thinking about this as well. Having to call a `to_bson()` function after calling `build()` didn't seem like the best idea. I was able to recover the manual implementation, so let me know how that looks!

Reply:
It looks like this is simply something `TypedBuilder` doesn't support yet; I'm fine with using the manual implementation, I don't think it'll matter to users one way or the other as long as how it works is consistent.